libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,221 @@
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    if (!isEnhancedInitialized) {
+      isEnhancedInitialized = true;
+      originalCallback = callback;
+      originalLooper = looper;
+      onFreeDrive();
+    } else {
+      locationEngine.requestLocationUpdates(request, callback, looper);
+    }
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
+    if (isConfigured.get()) {
+      removeLocationUpdates(callback);
+      if (future != null) {
+        future.cancel(false);
+        future = null;
+      }
+    }
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(originalLooper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+        @Override
+        public void run() {
+        if (rawLocation != null) {
+          final Location enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+              @Override
+              public void run() {
+              originalCallback.onSuccess(LocationEngineResult.create(enhancedLocation));
+            }
+          });
+        }
+      }
+    }, 1500, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  void removeOriginalLocationUpdates() {
+    if (isConfigured.get()) {
+      if (originalCallback != null) {
+        removeLocationUpdates(originalCallback);
+        originalCallback = null;
+      }
+      isEnhancedInitialized = false;
+    }
+  }
+
+  void onNavigationKilled() {
+    if (isConfigured.get()) {
+      if (future != null) {
+        removeLocationUpdates(callback);
+        future.cancel(false);
+        future = null;
+        executorService.shutdown();
+        removeOriginalLocationUpdates();
+      }
+    }
+  }
+
+  void configure(@NonNull File path, @NonNull String version, @NonNull String host,
```

Comment:
This operation and its results are relevant regardless of the location engine decorator instance. For example, now when new engine is set way after initialization of the navigation, the new instance of the wrapper will return `false` for `isConfigured`, while it's actually already configured.

Solutions I can see:
- make the decorator a singleton that has pluggable location engine instances
- move the nav-native configuration up, to `MapboxNavigation`
- invoke the configure call for each decorator instance again, whenever an instance of the engine is set in `MapboxNavigation` (is that call expensive?)